core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -378,7 +289,8 @@ public List<TableIdentifier> listTables(SessionContext context, Namespace ns) {
               paths.tables(ns),
               queryParams,
               ListTablesResponse.class,
-              headers(context),
+              ImmutableMap.of(),
+              authManager.contextualSession(context, catalogAuth),
```

Comment:
It feels like we're repeating this same pattern for all client calls now.  Wouldn't we be able to wrap the client so that all we need is the context here?

Reply:
Not sure I follow, can you give me an example of how you would see this call once refactored?

Reply:
(bear in mind this isn't fully fleshed out)

We're exposing a lot of the mechanics of how we do auth as opposed to forcing it through an API path that ensures use of the auth.  Consider an alternative where we have to get at the client through the auth manager/session:

```java
authManager.clientForSession(context, catalogAuth) // Returns a wrapped RESTClient
            .get(
              paths.tables(ns),
              queryParams,
              ListTablesResponse.class,
              configHeaders,
              ErrorHandlers.namespaceErrorHandler());
```

If requests are forced through the session, then it's more explicit.  It just feels like we're doing a lot of hand-crafting.

Reply:
I had initially something that resembles your suggestion. I discarded it for two reasons:

1. `RESTClient` implementations are stateful (and closeable); "wrapping" a REST client with the auth session may be error-prone: should implementors return a new instance, or the same one? Should callers close the wrapped client or not?
2. Wrapping makes the code slightly less testable: you don't "see" the authentication data anymore by simply inspecting the calls that the catalog made to the `RESTClient` API.

But I agree that the wrapping approach is less invasive, so let's give it another try. 

I think that to address concern #1 above, we can work around misbehaving implementations with good documentation. 

To address #2 above, we'd certainly need to revisit tests a little bit and introduce a better way to make verifications, ideally capturing which requests were made and what was the internal state of the REST client, by the time a given request was executed.